# Aufgabenblatt 4: Arrays

## Aufgabe 1: Wiederholung Strings
### a) nutue print für die folgenden Beispiele:
- referenziere auf einen `exemplarischen Ordner-Pfad` z.B: c:\ordner\datei

c:\ordner\datei


- erstelle einen String mit Sichtbaren Anführungszeichen

I'm lazy


- erstelle einen String über mehrere Zeilen (inkl. Zeilenumbruch)

Zeilen
umbruch


### b) kombiniere String mit Variablen
- gebe einen String mit einer Variable in einer print Funktion gemeinsam aus. z.B. `I'm 35`

I'm 35


- gibt es eine andere Schreibweise

I'm 35


### c) Verkettung und Bearbeitung von Strings
- erstelle einen String `string_1` mit dem Text `Python ist` mittels Anführungszeichen

'Python ist'

- erstelle einen String `string_2` mit dem Text `super` mittels einer Funkition

' super '

- erstelle einmal einen String `string_3` mit dem Text `:)` nach deiner Wahl

':)'

- füge die drei Strings zu einem String `string_all` zusammen

'Python ist super :)'

- schreibe mittels Multiplikator den folgenden `Python ist super super super :)` in die Variable `string_multi`

'Python ist super  super  super :)'

## Aufgaben 2: Bearbeiten von Strings
### a) Grundfunktionen

Analysiere den String ``string_find`` und bearbeite die nachfolgenden Schritte 

In [ ]:
# String für weitere Bearbeitung erstellt
string_find = "  THIS and that and borring stuff! bla "

- gebe die `Länge` des Strings `string_find` an
<br> Ergebnis 39

39


- gebe die `erste Position` des Buchstabens `a` aus dem String `string_find` an
<br> Ergebnis 7

7


- prüfe, ob der String `string_find` mit `tha` beginnt

False


- `veränadere` den Buchstaben `a` in `A` über alle Elemente in dem String `string_find`

  THIS And thAt And borring stuff! blA 


- prüfe, ob der String `string_find` mit den der String mit `ff` endet

False


- gebe den String `string_find` nur mit `klein Buchstaben` aus (lower case)

  this and that and borring stuff! bla 


- Zähle wie viele `a` es in dem String `string_find` gibt?
<br> Ergebnis 4

4


- gebe den String `string_find` nur mit `gross Buchstaben` aus

  THIS AND THAT AND BORRING STUFF! BLA 


- gebe die Stelle des `letztens` `a` innerhalb des Strings `string_find` an
<br> Ergebnis 37

37


- `enterferne alle Leerzeichen` aus dem String `string_find`

THISandthatandborringstuff!bla


- `lösche **nur** alle Leerzeichen am Anfang und am Ende` aus dem String `string_find` und 
- Zähle alle `Elemente` des Strings und übergebe das Ergebnis in eine Variable `string_find_len`.
- gebe den neuen `string_find` und das Ergebnis mittels print aus
<br> Ergebnis 36

THIS and that and borring stuff! bla 36


- nun `lösche nur die ersten Leerzeichen am Anfang` aus dem String `string_find` und 
- übergebe das Ergebnis, vom zählen aller Elemente in einer neuen Variable wieder. Wie ist die `Länge`?
<br> Ergebnis 38

38


### b) String zusammenführen

- Erstelle die beiden Strings jeweils in einer Variabeln
    - `string_1` = "Python is"
    - `string_2` = "amazing!"
- führe anschliessend die beiden Strings in eine Variable zusammen, sodass `Python is amazing!` herauskommt
- Gibt es mehr als eine Möglichkeit?


Python is amazing!
Python is amazing!


- verändere mittels einer `build in Funktion` in der Variable `string_2`. Es soll der Begriff `amazing` durch `booring`ausgetauscht werden
- gebe das Ergebnis mittels print aus

Python is borring!


#### c) Grundaufgaben Erweiterung

`string_find` (aus a) soll wie folgt bearbeitet werden. Die Bearbeitungen sollen in der Variable `liste_string_find` gespeichert werden:
- alle Leerzeichen am Ende und am Anfang aus `string_find` sollen enttfernt werden
- anschliessend sollen alle Buchstaben des Strings `string_find` **klein** geschrieben werden
- erstelle Listenelemente: d.h. jedes einzelne Wort aus dem String `string_find` soll ein eigenes Element in einer Liste werden
- stelle sicher, dass die Wörter in der Liste einmalig sind - d.h. ein Wort kann nicht mehrmals vor kommen
- Füge nun der Liste `liste_string_find` das Wort `is` hinzu. 
- Gebe die Liste mittles der print Funktion aus.
<br><br> **Ergebnis**: 'that', 'this', 'bla', 'borring', 'stuff!', 'and', 'is' -> Rheinfolge kann abweichen!!!

['that', 'this', 'bla', 'borring', 'stuff!', 'and', 'is']

- erstelle aus der Liste mittels Slicen den Satz: `this is borring` und gebe diesen aus

'this is borring'

- erstelle aus der Liste den Satz: `blablablablablabla` --> bla kommt 6 mal vor

'blablablablablabla'

- wie häufig ist `bla` in `blablablablablabla` enthalten --> lass den code zählen

6

### d) Was sind die Unterschiede bei split?
- Was ist der unterschied zwischen `split(sep = None)`, `split()` and `split(sep = " ")`?
<br> <br> Unteruche es an dem String `split_string`

In [ ]:
split_string = "String mit  unterschiedlichen   Anzahl an  Leerzeichen."

['String', 'mit', 'unterschiedlichen', 'Anzahl', 'an', 'Leerzeichen.']
['String', 'mit', 'unterschiedlichen', 'Anzahl', 'an', 'Leerzeichen.']
['String', 'mit', '', 'unterschiedlichen', '', '', 'Anzahl', 'an', '', 'Leerzeichen.']


### e) Build-In Function "format" 
#### bearbeiten von Strings mittels format. Diese Funktion ist z.B. sehr hilfreich im Zusammenhang bei `print`
Befülle den Text `"Mein Name ist ... und ich bin ... Jahre alt. Die Aufgaben ist ..."`
- einmal mit den Variablen name, age, value für die Punkte

Mein Name ist Kai und ich bin 35 Jahre alt. Die Aufgaben ist leicht.


- neben den Variabeln aus der Aufgabe davor, kann man format auch mit Positionsbestummungen (index) nutzen. Schreibe nun den Code von der AUfgabe davor um für die Indexbestimmung

Mein Name ist Kai und ich bin 35 Jahre alt. Die Aufgaben ist leicht.


- schriebe den Code von oben nun ohne Variabeln oder Index. **Tipp** Variabeln bezeichnungen sind direkt im der format Funktion 

Mein Name ist Kai und ich bin 35 Jahre alt. Die Aufgaben ist leicht.


 `Inhaltsfragen`
- Was passiert wenn die Reihnfolge im format jeweils für die drei Codearten oben geändert werden?
- Rheinfolge ist jetzt, `leicht, Kai, 35`

Mein Name ist Kai und ich bin 35 Jahre alt. Die Aufgaben ist leicht.
Mein Name ist leicht und ich bin Kai Jahre alt. Die Aufgaben ist 35.
Mein Name ist leicht und ich bin Kai Jahre alt. Die Aufgaben ist 35.


`Index` geht auch richtig - Wie?

Mein Name ist Kai und ich bin 35 Jahre alt. Die Aufgaben ist leicht.


 `Inhaltsfragen`
 - Was passiert wenn nicht alle Elemente übergeben werden?

### e) quote 007
- schreibe den Satz: `"My name is James, James Bond"` verwende hierfür **format** und gebe nur einen Parameter den Namen `James` wieder

My name is James, James Bond


### f) Aufgabe use: zip
- kombieniere die beiden nachfolgenden Listen mittels eines for-loops
- das Ergebnis soll u.a. sein: `Kai ist 36 Jahre alt`
<br><br> names_liste = ['Kai', 'James', 'Hulk']
<br> age_liste = [35, 42, '38']

Kai is 35 years old.
James is 42 years old.
Hulk is 38 years old.


### g) Strings justieren anhand der Länge
- Gebe die einzelnen Elemente der Liste `string_liste_just` aus? 

In [ ]:
string_liste_just = ["kurz", "mittel lang", "der längste String"]

kurz
mittel lang
der längste String


- Nun sollen die Elemente der `string_liste_just`  **`rechtsbündig`** ausgegeben werden (an dem maximalen Wert)
<br><br>
**Tipps:**
- kreiere eine Variable max_len
- du musst die Ausrichtung mitgeben werden --> Bestimme die maximale Länge der Elemente im Array --> mittels for-loop
- in einem zweiten foor-loop kannst du die justierung vornehmen mittels rjust

              kurz
       mittel lang
der längste String


### h) Integer justieren auf Punkt Ebene

In [51]:
punkt_daten = [3.21, 30.4, 2.575, 100.1001]
punkt_daten

[3.21, 30.4, 2.575, 100.1001]

- Gebe die einzelnen Elemente der Liste `string_liste_just` aus? 

3.21
30.4
2.575
100.1001


- gebe die Liste `punkt_daten` so aus, dass die Punkte auf der `gleichen höhe` sind
<br><br>
**Tipp**
- ähnlich zu der Aufgabe davor
- es müssen aber die Indexes oder länge von ELementen vor und nach dem Punkt bestimmt werden 

Elemente vor dem Punkt: 3 
Elemente nach dem Punkt: 4
  3.21  
 30.4   
  2.575 
100.1001


### i) Aufgabe: Operatoren 
- benutze Operatoren um zu prüfen, ob der Buchstabe `z` in dem Text aus der `Aufgabe e (Quote 007)` enthalten ist 

- ist der Buchstabe `d` in dem Text aus der `Aufgabe e (Quotw 007)` enthalten?

True

## Aufgabe 3: Listen bearbeiten
### a) Benutze verschiedene Funktionen um eine Liste zu bearbeiten

- erstelle zuerst eine leere Liste mit dem Namen `list_example`

[]

- füge der Liste den Wert `1` hinzu

[1]

- füge nun der Liste den Wert `3, 5, 7, 2, 13, 17` am Listenende hinzu --> also nach `1`

[1, 3, 5, 7, 2, 13, 17]

- füge der Liste den Wert `2` nach dem Wert `5` (bzw. an dem Index 2) hin zu

[1, 3, 2, 5, 7, 2, 13, 17]

- füge der Liste die folgenden Integerwerte `39, 27, 21` am Listenende hinzu

[1, 3, 2, 5, 7, 2, 13, 17, 39, 27, 21]

- entferne nur das `letzte` Element aus der Liste

[1, 3, 2, 5, 7, 2, 13, 17, 39, 27]

- entfenre das Element `7` aus der Liste

[1, 3, 2, 5, 2, 13, 17, 39, 27]

- entferne das `zweite Element` aus der Liste

[1, 3, 5, 2, 13, 17, 39, 27]

- sortiere die Liste (beginnend mit dem geringsten Wert)

[1, 2, 3, 5, 13, 17, 27, 39]

### b) Bilde die Potenz der Listenwerte 

- Bilde für jedes Element der Liste `basis_liste` die dazugehörige Potenz
- schriebe den neuen Wert in eine neue Liste
- Sortiere das Ergebnis (beginnend mit dem höchsten Wert)

In [ ]:
basis_liste = [3, 2, 9, 5, 7, 4, 1]
basis_liste

[3, 2, 9, 5, 7, 4, 1]

[81, 49, 25, 16, 9, 4, 1]


### c) Erstelle Spielpaarungen

- Es gibt die Mannschaften `Norwegen, Schweden, Finnland, Dänemark`
- Diese Mannschaften sollen jeweils gegen einander spielen.
- erstelle alle verschiedenen möglichen Spielpaarungen (**Tipp: enumerate**) und gebe diese anschliessend aus

[['Norwegen', 'Schweden'], ['Norwegen', 'Finnland'], ['Norwegen', 'Dänemark'], ['Schweden', 'Finnland'], ['Schweden', 'Dänemark'], ['Finnland', 'Dänemark']]


- gebe automatisiert mittels for Loop und den Variabeln `team1 vs team2` die zuvor erstellten Spielpaarungen aus

Norwegen vs Schweden
Norwegen vs Finnland
Norwegen vs Dänemark
Schweden vs Finnland
Schweden vs Dänemark
Finnland vs Dänemark


### d) Listen anpassen

- Die `liste_a` ist die Basis Liste
- Wenn der Wert von `liste_a` in Liste `liste_b` enthalten ist, soll der Wert aus `liste_b` direkt nach dem gleichen Element in `listen_a` angefügt werden
- Wenn nicht soll der Wert an das Ende der `liste_a` hinzugefügt werden.
<br><br>
**Tipp**
- Verwende ein for-loop und eine if-else-cause
<br><br> Die Lösung sieht wie folgt aus: *['a', 'a', 'a', 'c', 'c', 'e', 'e', 'g', 'g', 'f', 'f', 'd', 'd', 'h']*

In [ ]:
liste_a = ['a','c', 'e', 'g']
liste_b = ['c', 'f', 'd', 'h', 'a', 'e', 'd', 'a', 'f', 'g']

['a', 'a', 'a', 'c', 'c', 'e', 'e', 'g', 'g', 'f', 'f', 'd', 'd', 'h']


### e) Elementensuche bei einer bestimmten Bedingung in einer Liste

- Gebe alle Elemente der Liste `search_list` aus, welche das Teilwort / Buchstabenfolge `und` in ihrem Wort enthalten haben
- schreibe nur die Wörter, welche ein `und` enthalten haben in eine neue Liste

In [ ]:
search_list = ['Mund', 'Lörrach', 'Hund', 'DHBW', 'rund', 'bunt', 'und']
search_list

['Mund', 'Lörrach', 'Hund', 'DHBW', 'rund', 'bunt', 'und']

Mund
Hund
rund
und


['Mund', 'Hund', 'rund', 'und']

## Aufgabe 4: Dictionaries bearbeiten
### a) Nutze Funktionen um eine Dictionarie zu bearbeiten

- definiere ein Dictionarie mit dem Namen `work_dict` mit den `Schlüssel:Wert-Paaren` 
<br> **name** = Kai as *str* <br> **age** = 35 as *int* <br> **3** = 2.5 as *float*

{'name': 'Kai', 'age': 35, 3: 2.5}

- Gebe den Wert zu dem Schlüssel `age` aus

35

- füge das Element `uni=dhbw` dem Dictionarie hinzu

{'name': 'Kai', 'age': 35, 3: 2.5, 'uni': 'dhbw'}

- verändere die folgende Wert und bestimme die dazugehörigen Schlüssel:
    - `3` zu `1,6` **und**  
    - `uni` zu `DHBW Lörrrach` im dictionarie

{'name': 'Kai', 'age': 35, 3: 1.6, 'uni': 'DHBW Lörrrach'}

- entferne den Schlüssel `age` aus dem Dictionarie

{'name': 'Kai', 3: 1.6, 'uni': 'DHBW Lörrrach'}

- gebe alle Schlüssel aus dem Dictionarie wieder

dict_keys(['name', 3, 'uni'])

- gebe alle Werte aus dem Dictionarie wieder 

dict_values(['Kai', 1.6, 'DHBW Lörrrach'])

- gebe alle Schlüssel-Wert-Paar wieder

dict_items([('name', 'Kai'), (3, 1.6), ('uni', 'DHBW Lörrrach')])

- itteriere nun über alle Schlüsselelemente des Dictionaries und gebe die jeweiligen dazugehörigen Werte aus

Kai
1.6
DHBW Lörrrach


### b) Lohnverteilung für Postionen

- schreibe ein Dictionarie, indem der CEO = 10.000, der Tramleiter = 5.000, der Mitarbeiter = 2.5000 und der Student = 1250 verdient

{'CEO': 10000, 'Teamleiter': 5000, 'Mitarbeiter': 2500, 'Student': 1250}


- bestimme für die Position `Mitarbeiter` das Gehalt und gebe es als String folgendes mittels format aus: `Position` verdient `Summe` Wobei sowohl Position als auch Summe eine Variable sind

Mitarbeiter verdient 2500


- Aufgrund der derzeitig hohen Inflation sollen alle Gehälter um 9 % steigen. 
- Berechne die neuen Gehälter im Dictionarie und speicher den neuen Wert im Dictionarie
<br> `verwende eine for Schleife hier und nicht update`

{'CEO': 10900.0,
 'Teamleiter': 5450.0,
 'Mitarbeiter': 2725.0,
 'Student': 1362.5}

## Aufgabe 5: Bearbeiten von Mengen
### a) Funktionen um eine Set zu bearbeiten
- definiere ein **Set** `set_one` mit den Werten `1, 2, 3, 1, 4, 3, 2, 8`

{1, 2, 3, 4, 8}

- definiere nun ein weiteres **Set** `set_two` mit den Werten `1, 1, 6, 6, 4, 3, 5, 7`

{1, 3, 4, 5, 6, 7}

- füge nun `set_one` den Wert `5` hinzu

{1, 2, 3, 4, 5, 8}

- entferne anschliessend aus dem `set_one` den Wert `2`

{1, 3, 4, 5, 8}

- entferne aus `set_one` den Wert `8` mit einer **anderen** Funktion

{1, 3, 4, 5}

- gebe nun die beiden sets einmal aus

{1, 3, 4, 5}
{1, 3, 4, 5, 6, 7}


- verbinde die beiden Sets `set_one und set_two` zu einem gemeinsamen Set `set_union`

{1, 3, 4, 5, 6, 7}

- Bestimme nun die Gemeinsamenwerte aus `set_one und set_two`

{1, 3, 4, 5}

- gebe nun die Werte aus, welche nicht in beiden sets vorkommen
- Die Basis für die nicht Gemeinsamkeiten soll `set_two` sein

{6, 7}

- Überprüfe, ob es eine leere Schnittmenge gibt

False

- Überprüfe, ob `set_one` eine Untermenge von `set_two` ist

True

- Überprüfe, ob `set_one` eine Obermenge von `set_two` ist

False